README.md
[![npm](https://img.shields.io/npm/v/webcrypto-crypt.svg)]() [![Travis](https://img.shields.io/travis/c2fo-lab/webcrypto-crypt.svg)]() [![license](https://img.shields.io/github/license/c2fo-lab/webcrypto-crypt.svg)]() [![npm](https://img.shields.io/npm/dw/webcrypto-crypt.svg)]() [![node](https://img.shields.io/node/v/webcrypto-crypt.svg)]() [![GitHub last commit](https://img.shields.io/github/last-commit/c2fo-lab/webcrypto-crypt.svg)]()

  * [Introduction](#introduction)
  * [Install](#install)
  * [Quickstart](#quickstart)
  * [Examples](#examples)
  * [Test](#test)
  * [API](#api)
  * [Processing node streams](#processing-node-streams)
  * [Command-line](#command-line)
  * [Header structure](#header-structure)
  * [Security](#security)
  * [See also](#see-also)
  
# Introduction

**webcrypto-crypt** encrypts and decrypts [data at rest](https://en.wikipedia.org/wiki/Data_at_rest#Concerns_about_data_at_rest) in node or the browser using [secret key cryptography](https://www.ibm.com/support/knowledgecenter/SSYKE2_8.0.0/com.ibm.java.security.component.80.doc/security-component/jsse2Docs/secretkeycryptography.html) in conjunction with a [passphrase](https://en.wikipedia.org/wiki/PBKDF2#Purpose_and_operation).

The package will use either the [Window.crypto](https://developer.mozilla.org/en-US/docs/Web/API/Window/crypto) or [node-webcrypto-ossl](https://github.com/PeculiarVentures/node-webcrypto-ossl) library, depending on where it's running.

Encryption and decryption depend on the [key derivation function](https://en.wikipedia.org/wiki/Key_derivation_function) referred to as [PBKDF2](https://www.w3.org/TR/WebCryptoAPI/#pbkdf2) and the      [symmetric key algorithm](https://en.wikipedia.org/wiki/Symmetric-key_algorithm) called [AES-GCM](https://www.w3.org/TR/WebCryptoAPI/#aes-gcm).

For node, the package provides streaming methods and command-line utility **wcrypt** to deal with files and pipes.

# Install

    λ npm install -g webcrypto-crypt

# Quickstart

## Command-line

### From prompt

    λ wcrypt | wcrypt -d
    Data to encrypt: no honour among consultants.
    Passphrase?
    Confirm passphrase:
    Passphrase?
    no honour among consultants.λ

### From and to file

    λ wcrypt -i LICENSE -o LICENSE.wcrypt
    Passphrase?
    Confirm passphrase:
    λ wcrypt -d -i LICENSE.wcrypt
    Passphrase?
    MIT License

    Copyright (c) 2017 C2FO...

### From arg
 
    λ wcrypt -a 'edge of a dynastic rebellion' > mydata.wcrypt
    Passphrase?
    Confirm passphrase:
    λ cat mydata.wcrypt | wcrypt -d  # or wcrypt -di mydata.wcrypt
    Passphrase?
    edge of a dynastic rebellionλ

### From pipe

    λ echo 'pretty despite its implicit threat' | wcrypt | wcrypt -d
    Passphrase?
    Confirm passphrase:
    Passphrase?
    pretty despite its implicit threat

## Node.js

```javascript
    const Wcrypt = require('webcrypto-crypt')
    const wcrypt = new Wcrypt.Cipher('justtesting')
    wcrypt.encrypt('no honour among consultants.')
    .then((buf) => {
      // do something with buf
    })
```

## Browser

```jsx
  <html>
    <head>
      <script src="dist/wcrypt.js"></script>
      <script>
        var wcrypt = new Wcrypt.Cipher('justtesting');
        wcrypt.encrypt('edge of a dynastic rebellion')
          .then((buf) => {
            console.log('Results %s', buf);
          });
      </script>
    </head>
    <body>
    </body>
  </html>
```

# Examples

## Node.js

```javascript
#!/bin/sh
':' //; exec "$(command -v nodejs || command -v node)" "$0" "$@"
'use strict';
const readlineSync = require('readline-sync')
const Wcrypt = require('webcrypto-crypt')

function askForData () {
  return readlineSync.question('Data to encrypt? ')
}

function askForPassphrase () {
  return readlineSync
    .question('Passphrase to use? ',
      { hideEchoBack: true, mask: '' }
    )
}

var wcrypt = new Wcrypt.Cipher(askForPassphrase())

wcrypt.encrypt(askForData())
  .then((data) => {
    console.log(`
      Encrypted, hex-encoded: ${data.toString('hex')},
      Encrypted, base64-encoded: ${data.toString('base64')},
      Encrypted, web-safe base64-encoded: ${wcrypt.uriSafeBase64(data)},
      Decrypted using same passphrase:`
    )
    wcrypt.decrypt(data)
      .then((data) => {
        console.log('        ' + Buffer.from(data).toString('utf8'))
      })
  })
```

    λ node examples/prompts.js
    Passphrase to use?
    Data to encrypt? ambiguous or even contradictory situations

    Encrypted, hex-encoded: c0e82e52c7400925f43e2753249afd8b1527a193917267df53d73d63f287519717c73d3e6c1a3da1797ef19055a49b8402cbb0e10f0a5a294545
    Encrypted, base64-encoded: wOguUsdACSX0PidTJJr9ixUnoZORcmffU9c9Y/KHUZcXxz0+bBo9oXl+8ZBVpJuEAsuw4Q8KWilFRQ==
    Encrypted, web-safe base64-encoded: wOguUsdACSX0PidTJJr9ixUnoZORcmffU9c9Y_KHUZcXxz0.bBo9oXl.8ZBVpJuEAsuw4Q8KWilFRQ--

    Decrypted using same passphrase:
    ambiguous or even contradictory situations
    λ

## Browser

```jsx
<html>
  <head>
    <script src="dist/wcrypt.js"></script>
    <script>
      var wcrypt = new Wcrypt.Cipher(prompt("Secret? "));
      wcrypt.encrypt(prompt("Data to encrypt? "))
        .then((data) => {
          console.log(`
            Encrypted, hex-encoded: ${ data.toString("hex") },
            Encrypted, base64-encoded: ${ data.toString("base64") },
            Encrypted, web-safe base64-encoded: ${ wcrypt.uriSafeBase64(data) },
            Decrypted using same passphrase:`
          );
          wcrypt.decrypt(data)
            .then((data) => {
              console.log('         ' + Buffer.from(data).toString("utf8"));
            });
        });
   </script>
 </head>
 <body>
   (Check the Developer console.)
 </body>
</html>
```
    
    // In the dev console, after calls to window.prompt() are satisfied, e.g.:
    1GTLSRGmdkiAn6pwBB1XWYllS2VgF87h8575Ok2NlIzdH42YgoUFRpFiqbYqj2BKLJEQqoQ-
    index.html:11 in a world that allowed such mistakes

# Test

## Tested environments

| **OS** | **Environment** | **Version** |
| :-------- | :------- | :------- |
| Mac El Capitan | Chrome  | 79.0.3945.130 |
| Mac El Capitan | Firefox | 72.0.2  |
| Mac El Capitan | Node    | v13.9.0 |
| Mac El Capitan | Safari  | 11.1.2  |

## Node.js

    λ npm run test

## Browsers

A file that should run the module tests in the browser is ```test/browser/wcrypt-test.html```. webcrypto-crypt currently injects [browserify](https://www.npmjs.com/package/buffer)'s ```Buffer``` as a global object.

# API

## Wcrypt.getSignature()

Return the current file signature in use by this library as a ```Buffer```.

## Wcrypt.parseHeader(Buffer data)

Provided ```data``` is a valid webcrypto-wcrypt header, parse it and return an object with the following structure:

```json
    {
      "material": {
        "iv": "<iv>",
        "salt": "<salt>",
      },
      "config": {
        "crypto": {
          "tagLength": "<tagLength>"
        },
        "derive": {
          "iterations": "<iterations>"
        }
      }
    }
```

## Wcrypt.setHeader(Object config)

Applies this configuration object to the current file header definition.

## Wcrypt.DEBUG = true | false

If set to ```true```, send debugging statements to stderr.  Default ```false```.

## new Wcrypt.Cipher(String passphrase || Object options)

Instantiate a webcrypto-crypt object using just a passphrase or more options beyond the passphrase.  When passing in an object, the minimum specification looks like ```{material: { passphrase: <your passphrase> } }```.  Possible options are described below:

```javascript
  var wcrypt = new Wcrypt.Cipher({
    config: {
      crypto: {
        usages: [myU1, myU2...],  // default ['encrypt', 'decrypt']
        tagLength: myTagLength,   // default 128
      },
      delimiter: myDelimiter,     // default '<WcRyP+>'
      derive: {
        hash: myHashFunction,     // default 'SHA-512'
        iterations: myIterations, // default 2000
        length: myLength,         // default 128
      },
      paranoid: true              // check each encrypted block
                                  // for incidental presence of
                                  // the delimiter, re-encrypt data
                                  // if detected (very slow).
                                  // default false
    },
    material: {
      iv: myInitializationVector, // default getRandomValues(new Uint8Array(12))
      passphrase: 'justtesting',  // REQUIRED, passphrase as String
      salt: mySalt                // default getRandomValues(new Uint8Array(16))
    }
  })
  .then(...
```

Callers passing in their own value for initialization vector (```material.iv```) must ensure that it is [never reused under the same key](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Initialization_vector_.28IV.29).  Likewise, callers passing in their own value for key derivation salt (```material.salt```) should ensure it contains data that explicitly [distinguishes between different operations](https://www.rfc-editor.org/rfc/rfc8018.txt).

## wcrypt.createHeader()

Return a ```Buffer``` filled with the appropriate seed data for encryption.  See [these lines](https://github.com/c2fo-lab/webcrypto-crypt/blob/master/index.js#L293-L300).

## wcrypt.decrypt(Buffer data)

Decrypt ```data``` by first parsing its header section to extract ```config``` and ```material``` settings.  The innvoking ```wcrypt``` object will assume any extracted ```config``` and ```material``` settings for all subsequent operations until overwritten n by reading a different header.  Returns a promise with its resolved value set to a Buffer of the decrypted data.

## wcrypt.delimiter

The current delimiter in use by this library, e.g., ```<WcRyP+>```.

## wcrypt.encrypt(Buffer data)

Encrypt the given data and include a header.  Returns a promise with its resolved value set to a Buffer of the encrypted data.

## wcrypt.encryptDelimited(Buffer data)

Encrypt the given data and include a delimiter.  Returns a promise with its resolved value set to a Buffer of the encrypted data.

## wcrypt.getDelimiter()

Return the delimiter webcrypto-crypt is currently configured to use as a ```Buffer```.

## wcrypt.name

The current name of this library, e.g., ```webcrypto-crypt```.

## wcrypt.rawDecrypt(Buffer data, Object options)

Decrypt ```data```.  Assumes no header present unless ```assumeHeader: true``` is passed in via an ```options``` object.  Returns a promise with its resolved value set to a Buffer of the decrypted data.

## wcrypt.rawEncrypt(Buffer data, Object options)

Returns a promise with its resolved value set to a Buffer of the encrypted data.  The first argument is the data to encrypt.  The second argument ```options``` is optional and currently supports two attributes:

    wcrypt.rawEncrypt('Some text to encrypt', {appendDelimiter: true});

Will append the ```wcrypt.delimiter``` to the end of the encrypted data before returning it.  And:

    wcrypt.rawEncrypt('Some text to encrypt', {includeHeader: true});

Will append a cleartext header to the returned data including: ```initialization vector```, ```iterations```, ```salt```, and ```tagLength``` used in the encryption.  See [these lines](https://github.com/c2fo-lab/webcrypto-crypt/blob/master/index.js#L293-L300).

## wcrypt.subtle

The [SubtleCrypto](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto) object currently in use.

## wcrypt.uriSafeBase64(Buffer data)

Convenience method for browser contexts to encode the passed-in ```data``` using [web safe base 64 alphabet](https://tools.ietf.org/html/rfc4648#section-5) and return the result.

## wcrypt.version

The current version of this library, e.g., ```0.1.2```.

# Processing node streams

In a node.js environment, ```webcrypto-crypt/lib/node-streams``` provides convenience methods for encrypting and decrypting [Readable Streams](https://nodejs.org/api/stream.html#stream_readable_streams).  Please assume the following lines precede these examples:

```javascript
        const fs = require('fs')
        const Wcrypt = require('webcrypto-crypt')
        const WcryptStream = require('webcrypto-crypt/lib/node-streams')
        const cleartext = 'LICENSE'
        const ciphertext = cleartext + '.wcrypt'
```

## Encrypt a readable stream

When encrypting you need to pass in your own ```wcrypt``` object:

```javascript
        const wcrypt = new Wcrypt.Cipher('justtesting')
        const clear = fs.createReadStream(cleartext)
        const encrypted = fs.createWriteStream(ciphertext)
        const read = WcryptStream.encrypt(wcrypt, clear)

        read.on('data', (chunk) => {
          encrypted.write(chunk)
        })
        read.on('end', () => {
          console.error('wrote ' + ciphertext)
        })
```

## Decrypt a readable stream

When decrypting you need to pass in the passphrase:

```javascript
        const decrypt = fs.createReadStream(ciphertext)
        const read = WcryptStream.decrypt('justtesting', decrypt)
        read.on('data', (chunk) => {
          console.log(chunk.toString('utf8'))
        })
```

# Command-line

Installing **webcrypto-crypt** also installs a command-line utilty, **wcrypt**:

## Help

### TTY

    λ wcrypt -h
    Usage: wcrypt [options]
    
    Options:
      --debug, -D    write debug info to stderr         [boolean] [default: "false"]
      --help, -h     Show help                                             [boolean]
      --outfile, -o  write data to this file                                [string]
      --infile, -i   read data from this file                               [string]
      --arg, -a      read data from command line                            [string]
      --decrypt, -d                                       [boolean] [default: false]
    
    Examples:
      wcrypt -i msg.txt -o msg.wcrypt      file to encrypt
      wcrypt -di msg.wcrypt -o msg.txt     file to decrypt
      wcrypt -a "nonessential appliances"  string to encrypt

### Piped

    λ echo '' | wcrypt -h
    Usage: data | wcrypt [options]
    
    Options:
      --debug, -D    write debug info to stderr         [boolean] [default: "false"]
      --help, -h     Show help                                             [boolean]
      --outfile, -o  write data to this file                                [string]
      --decrypt, -d                                       [boolean] [default: false]
    
    Examples:
      cat msg.txt | wcrypt -o msg.wcrypt       encrypt,store in file
      cat msg.wcrypt | wcrypt -d               decrypt
      curl http:..:4196 | wcrypt > aud.wcrypt  stream to encrypt

## Examples

Note that some of these examples may take a few minutes to download or stream.

Note also that you may set the environment variable **WCRYPT_PASSFILE** to have **wcrypt** skip its passphrase prompts.

### Text file

    λ export TXTFILE_URL="http://web.archive.org/web/20160514230001/http://www.mit.edu/~yandros/doc/command.txt"
    λ curl -Ss "$TXTFILE_URL" | wcrypt > command.wcrypt
    Passphrase?
    Confirm passphrase:
    λ wcrypt -di command.wcrypt -o command.txt
    Passphrase?
    λ less command.txt

### Audio file

    λ export AUDIO_URL="https://ia802509.us.archive.org/18/items/interview-with-neal-stephenson.JUcDkK.popuparchive.org/focus990527b.mp3"
    λ curl -NSs "$AUDIO_URL" | wcrypt > ns-interview.wcrypt # may take a few minutes...
    Passphrase?
    Confirm passphrase:
    λ wcrypt -di ns-interview.wcrypt -o ns-interview.mp3
    Passphrase?
    λ open ns-interview.mp3

### Audio stream

    λ export STREAM_URL="http://vprbbc.streamguys.net/vprbbc24-mobile.mp3"
    λ curl -NSsm 10 "$STREAM_URL" | wcrypt | wcrypt -d | mpg123 -q -
    Passphrase?
    Confirm passphrase:
    Passphrase?

### .wcryptpass

If no passphrase is specified, ```wcrypt``` will consult the environment variable ```WCRYPT_PASSFILE``` and try to use its contents as the path to a file containing the passphrase.  If ```WCRYPT_PASSFILE``` is not defined, ```wcrypt``` will check for a ```.wcryptpass``` file in the user's home directory for the same purpose.  In either case, on Unix systems the permissions of this file need to disallow access to world or group, achievable with, e.g., ```chmod 0600 ~/.wcryptpass```. If the permissions are less strict, the file will be ignored.

# Header structure

See [these lines](https://github.com/c2fo-lab/webcrypto-crypt/blob/master/index.js#L322-L329).

# Security

Security Mail: labs@c2fo.com<br>
PGP key fingerprint: ````E838 B51C C63F 7ED6 0980 9535 4D46 5218 A674 6F81````<br>
Keyserver: pgp.mit.edu<br>

# See also

* [ccrypt](http://ccrypt.sourceforge.net/)
* [introduction to cryptography](https://www.nrc.gov/site-help/e-submittals/intro-crypt.html)
* [webcrypto-examples](https://github.com/diafygi/webcrypto-examples)

# To do
* Promisify mocha tests
* Improve stream handling
* Add key usage permissions to file header


CHANGES.md
## 2020-07-10  Release 0.2.6.

Bump .travis.yml node version from 12.4.0 to 13.9.0

## 2020-07-08  Release 0.2.5.

Upgrade misc npm packages.
Change node-webcrypto-ossl constructor call to conform to latest API

## 2020-03-14  Release 0.2.4

Bump [acorn](https://github.com/acornjs/acorn) from 7.1.0 to 7.1.1.
Upgrade misc npm packages.

## 2020-01-28  Release 0.2.3

Update code snippets in README.

## 2020-01-28  Release 0.2.2

Fix link to lines showing header
Add back .eslintrc.js
Edit .travis.yml to request Node 12.4.0

## 2020-01-27  Release 0.2.1

Wcrypt.cipher() -> Wcrypt.Cipher() in README
Fix bin/wcrypt.js so it behaves as expected.

## 2020-01-27  Release 0.2.0

Comply with eslint-config-standard
Change Wcrypt.cipher() to Wcrypt.Cipher() to comply with eslint-config-standard
Upgrade NPM packages.
Modify browserify script & tests to work/pass with updated NPM packages.
Update README to reflect latest node and browser tests completed.

## 2018-11-26  Release 0.1.19

Address npm security advisory #720 (cryptiles)

## 2018-06-28  Release 0.1.18

Upgrade npm packages.

## 2017-11-14  Release 0.1.17

Update descriptive data in package.json.

## 2017-11-13  Release 0.1.16

Add use strict.
Update README.md.
Use backticks for interpolation in examples.

## 2017-11-13  Release 0.1.15

Add badges to README.md.
Fix github repo links in README.md.
Add .travis.yml.

## 2017-11-11  Release 0.1.14

Add 'paranoid' option to optionally prevent delimiter from showing up in encrypted blocks.
Add event name to emit() call.
Clean up error handling in promises.
Add warnings to README.md about unique initialization vector and salt.
Clean up text in README.md.

## 2017-08-31  Release 0.1.12

  Reduce number of errors emitted during stream error.
  Correct variable name in transcoder.
  Update README.md.

## 2017-08-01  Release 0.1.11

Fix error handling in node-streams.js.
Remove support for WCRYPT_PASS.
Add support for WCRYPT_PASSFILE.
Remove extraneous msgs from browserify script.
Clean up stdout messages in wcrypt cmdline tool.
Remove extraneous debug msgs.
Corrent some indentation.
Update README.md.

## 2017-07-25  Release 0.1.10

Make header definition dynamic based on config.
Add setHeader() class method.
Make getSignature() a class method.
Change file header to handle larger version numbers.
Remove extraneous pkg deps from wcrypt.js.

## 2017-07-13  Release 0.1.9

Update test data for previously failing test.

## 2017-07-13  Release 0.1.8

Add missing signature to SIGNED.txt.

## 2017-07-13  Release 0.1.7

Add PBKDF2 hash to file header.
Add missing catch clause to node-stream.
Catch errors in pipe() when streaming data.
Fix indentation.

## 2017-07-06  Release 0.1.6

Remove algorithm choice from custom config options.
Add derived key length config option.
Fix facility for passing in custom config to constructor.

## 2017-07-02  Release 0.1.5

Add helper lib for node streaming.
Make cmdline utility use new helper lib.
Move debug to class method.
Add test to check debug is working.
Fix indentation.

## 2017-06-04  Release 0.1.4

Remove extraneous pkgs from package.json.
Add package-lock.json.
Update README.md with Node 8 test.

## 2017-05-31  Release 0.1.3

Fix link to github repo.

## 2017-05-31  Release 0.1.2

Fix package versioning.

## 2017-05-31  Release 0.1.1

Fix decrypt() bug resetting material values.

## 2017-05-30  Release 0.1.0

Initial release.


test/debug.js
describe('Debug', function () {
  it('Sends output to stderr', function (done) {
    var exec = require('child_process').exec
    var source = 'node -e \'var Wcrypt = require (__dirname);' +
                'Wcrypt.DEBUG = true; var wcrypt = new Wcrypt.Cipher("123");\''
    exec(source, function (error, stdout, stderr) {
      if (error) {
        console.error(error)
      }
      if (stderr.match(/\[debug\] /)) { done() } else { done(new Error('No stderr output detected.')) }
    })
  })
})


test/mocks.js
var data = require('./data.js')
var fs = require('fs')
var imageClear = data.binary.encodings.plaintext.b64
var imageCipher = data.binary.encodings.ciphertext.b64
var imageFile = data.binary.plaintext
var toBuffer = require('blob-to-buffer')
var Transcoder = require('../lib/transcoder.js')
var transcoder = new Transcoder()

module.exports = function () {
  var mocks = {
    pngHash: '5795ae575f2a04c03ffd659dbe665cfad694f15f537b9f8b3673b3003a36da5d',
    pngCipherHash: '5041b67bf2aca626627a505943dc1dacc1ed278f659afe759de069b4fac0e346'
  }

  try {
    mocks.png = fs.readFileSync(imageFile)
    mocks.pngHash = '5795ae575f2a04c03ffd659dbe665cfad694f15f537b9f8b3673b3003a36da5d'
    mocks.pngCipher = Buffer.from(imageCipher, 'base64')
    mocks.pngCipherHash = '5041b67bf2aca626627a505943dc1dacc1ed278f659afe759de069b4fac0e346'
  } catch (err) {
    try {
      mocks.png = new Blob([transcoder.b642uint8(imageClear)], { type: 'image/png' })
      toBuffer(mocks.png, function (err, buffer) {
        if (err) throw err
        mocks.png = buffer
        mocks.pngCipher = Buffer.from(imageCipher, 'base64')
      })
    } catch (err) {
      console.error('XERROR: ' + err)
    }
  }

  mocks.altHash = 'SHA-256'
  mocks.altHashInvalid = 'SHA-DOG'
  mocks.altIter = 10000
  mocks.altIterInvalid = 'lorem ipsum'
  mocks.altLength = 128
  mocks.altLengthInvalid = 999
  mocks.altTagLength = 112
  mocks.altTagLengthInvalid = 999
  mocks.data = data
  mocks.iv = transcoder.str2ab('123456789012')
  mocks.passphrase = 'testSecret'
  mocks.salt = transcoder.str2ab('1234567890123456')
  mocks.altUsagesDecrypt = ['decrypt']
  mocks.altUsagesEncrypt = ['encrypt']
  mocks.altUsagesEncryptInvalid = ['decrypt']
  mocks.altUsagesDecryptInvalid = ['encrypt']

  return mocks
}


examples/prompts.js
#!/bin/sh
':' // ; exec "$(command -v nodejs || command -v node)" "$0" "$@"
'use strict'
const readlineSync = require('readline-sync')
const Wcrypt = require('..')

function askForData () {
  return readlineSync
    .question('Data to encrypt? ')
}

function askForPassphrase () {
  return readlineSync
    .question('Passphrase to use? ',
      { hideEchoBack: true, mask: '' }
    )
}

var wcrypt = new Wcrypt.Cipher(askForPassphrase())

wcrypt.encrypt(askForData())
  .then((data) => {
    console.log(`
        Encrypted, hex-encoded: ${data.toString('hex')},
        Encrypted, base64-encoded: ${data.toString('base64')},
        Encrypted, web-safe base64-encoded: ${wcrypt.uriSafeBase64(data)},

        Decrypted using same passphrase:`
    )
    wcrypt.decrypt(data)
      .then((data) => {
        console.log('        ' + Buffer.from(data).toString('utf8'))
      })
  })


index.js
'use strict'
const bindexOf = require('bindexof')
const Config = require('./lib/config.json')
const Package = require('./package.json')
const Transcoder = require('./lib/transcoder.js')
const Webcrypto = require('node-webcrypto-ossl')
const transcoder = new Transcoder()

global.Buffer = global.Buffer || Buffer

module.exports = exports = {

  Cipher: function (options) {
    var config = JSON.parse(JSON.stringify(Config))
    var material = {}
    var wcrypt = this

    try {
      wcrypt.crypto = new Webcrypto.Crypto()
    } catch (err) {
      wcrypt.crypto = window.crypto || window.msCrypto
    }

    if (typeof options === 'string') {
      options = {
        config: {},
        material: {
          passphrase: options
        }
      }
    } else if (Array.isArray(options) || typeof options === 'number') {
      throw Error(Config.err.passReq)
    } else if (!options.config) {
      options.config = {}
    } else if (!options.material) {
      options.material = {}
    }

    _overrideConfig(options.config)
    _setMaterial(options.material)
    exports.setHeader(config)

    exports.debug('Debugging enabled.')

    wcrypt.createHeader = function () {
      return Buffer.concat([
        exports.getSignature(config),
        Buffer.from(config.derive.iterations.toString(), 'utf8'),
        Buffer.from(config.crypto.tagLength.toString(), 'utf8'),
        Buffer.from(config.derive.length.toString(), 'utf8'),
        Buffer.from(config.derive.hash.toString(), 'utf8'),
        transcoder.ab2buf(material.iv),
        transcoder.ab2buf(material.salt),
        wcrypt.getDelimiter()
      ])
    }
    wcrypt.decrypt = function (data) {
      return wcrypt.rawDecrypt(data, { assumeHeader: true })
    }
    wcrypt.delimiter = config.delimiter
    wcrypt.encrypt = function (data) {
      return wcrypt.rawEncrypt(data, { includeHeader: true })
    }
    wcrypt.encryptDelimited = function (data) {
      return wcrypt.rawEncrypt(data, { appendDelimiter: true })
    }
    wcrypt.getDelimiter = function () {
      exports.debug('getDelimiter', config.delimiter)
      return Buffer.from(config.delimiter)
    }
    wcrypt.name = Package.name
    wcrypt.rawDecrypt = _decrypt
    wcrypt.rawEncrypt = _encrypt
    wcrypt.subtle = wcrypt.crypto.subtle || wcrypt.crypto.webkitSubtle
    wcrypt.uriSafeBase64 = function (data) {
      return transcoder.b642uri(
        transcoder.buf2b64(data)
      )
    }

    function _decrypt (chunk, options) {
      options = options || {}
      var data
      if (options.assumeHeader) {
        var parsed = exports.parseHeader(chunk)
        data = parsed.payload
        _overrideConfig(parsed.config)
        _setMaterial(parsed.material)
        exports.debug('_decrypt salt', Buffer.from(material.salt).toString('hex'))
      } else {
        data = chunk
      }
      exports.debug('_decrypt', JSON.stringify(options))
      return _getKey()
        .then(key => {
          material.key = key
          return _wDecrypt(data)
        })
        .catch(err => { throw err })
        .then(data => { return data })
        .catch(err => { throw err })
    }

    function _deriveKey () {
      exports.debug('_deriveKey salt', Buffer.from(material.salt).toString('hex'))
      return wcrypt.subtle.deriveKey(
        {
          name: config.derive.algorithm,
          salt: material.salt,
          iterations: config.derive.iterations,
          hash: config.derive.hash
        },
        material.baseKey,
        {
          name: config.crypto.algorithm,
          length: config.derive.length
        },
        true,
        config.crypto.usages
      )
    }

    function _encrypt (data, options) {
      options = options || {}
      exports.debug('_encrypt', JSON.stringify(options))
      if (typeof data === 'string') {
        data = Buffer.from(data, 'utf8')
      }
      return _getKey()
        .then(key => {
          material.key = key
          return _wEncrypt(data)
        })
        .catch(err => { throw err })
        .then(result => {
          if (config.paranoid) {
            if (bindexOf(result, wcrypt.getDelimiter())) { _encrypt(data, options) }
          }
          return _encryptedBlock(result, options)
        })
        .catch(err => { throw err })
    }

    function _encryptedBlock (data, options) {
      options = options || {}
      exports.debug('_encryptedBlock', JSON.stringify(options))
      if (options.appendDelimiter && !options.includeHeader) {
        return Buffer.concat([
          data,
          wcrypt.getDelimiter()
        ])
      } else if (!options.appendDelimiter && options.includeHeader) {
        return Buffer.concat([
          wcrypt.createHeader(),
          data
        ])
      } else if (options.includeHeader && options.appendDelimiter) {
        return Buffer.concat([
          wcrypt.createHeader(),
          data,
          wcrypt.getDelimiter()
        ])
      } else {
        return data
      }
    }

    function _getKey () {
      if (material.key) {
        return new Promise((resolve, reject) => {
          resolve(material.key)
        })
      } else {
        return _importKey()
          .catch(err => { throw err })
          .then(key => {
            material.baseKey = key
            return _deriveKey()
          })
          .catch(err => { throw err })
      }
    }

    function _getRandomBytes (numBytes) {
      var buf = new Uint8Array(numBytes)
      return wcrypt.crypto.getRandomValues(buf)
    }

    function _importKey () {
      if (typeof material.passphrase !== 'string') {
        return Promise.reject(config.err.passString)
      } else {
        return wcrypt.subtle.importKey(
          'raw',
          transcoder.str2ab(material.passphrase),
          { name: config.derive.algorithm },
          false,
          ['deriveKey']
        )
      }
    }

    function _overrideConfig (overrides) {
      overrides = overrides || {}

      if (!overrides.crypto) { overrides.crypto = {} }
      if (!overrides.derive) { overrides.derive = {} }

      exports.debug('_overrideConfig overrides', JSON.stringify(overrides))
      config.crypto.usages = overrides.crypto.usages || config.crypto.usages
      config.crypto.tagLength = overrides.crypto.tagLength || config.crypto.tagLength
      config.derive.algorithm = overrides.derive.algorithm || config.derive.algorithm
      config.derive.hash = overrides.derive.hash || config.derive.hash
      config.derive.iterations = overrides.derive.iterations || config.derive.iterations
      config.derive.length = overrides.derive.length || config.derive.length
      config.paranoid = overrides.paranoid || false
      exports.debug('_overrideConfig result', JSON.stringify(config))
    }

    function _setMaterial (params) {
      params = params || {}
      exports.debug('_setMaterial params', JSON.stringify(params))
      material.iv = params.iv || material.iv || _getRandomBytes(12)
      material.passphrase = params.passphrase || material.passphrase
      material.salt = params.salt || material.salt || _getRandomBytes(16)
      exports.debug('_setMaterial result', JSON.stringify(material))
    }

    function _wDecrypt (data) {
      return wcrypt.subtle.decrypt(
        {
          name: config.crypto.algorithm,
          iv: material.iv,
          tagLength: config.crypto.tagLength
        },
        material.key,
        transcoder.buf2ab(data)
      )
        .then(result => {
          return transcoder.ab2buf(result)
        })
        .catch(err => { throw err })
    }

    function _wEncrypt (data) {
      return wcrypt.subtle.encrypt(
        {
          name: config.crypto.algorithm,
          iv: material.iv,
          tagLength: config.crypto.tagLength
        },
        material.key,
        transcoder.buf2ab(data)
      )
        .then(result => { return transcoder.ab2buf(result) })
        .catch(err => { throw err })
    }
  },

  debug: function () {
    var msg = Array.prototype.slice.call(arguments)
    msg.unshift('[debug] ')
    msg = msg.join(' ')
    if (exports.DEBUG) { console.error(msg) }
  },

  delimiter: Config.delimiter,

  getSignature: function (config) {
    var v = (exports.version).split('.')
    for (var i = 0; i < v.length; i++) {
      if (v[i].length < 2) v[i] = '0' + v[i]
    }
    v = v.join('.')
    var signature
    if (config) { signature = config.signaturePrefix } else { signature = Config.signaturePrefix }
    signature = signature + v
    return Buffer.from(signature)
  },

  parseHeader: function (data) {
    const config = JSON.parse(JSON.stringify(Config))
    const material = {}

    exports.setHeader(config)

    var signature = (data.slice(0, exports.head.sig.e)).toString('utf8')
    var prefix = signature.substring(0, exports.head.pref)
    var version = signature.substring(exports.head.pref, exports.head.sig.e)

    if (prefix !== config.signaturePrefix) { throw new Error(Config.err.sigInvalid + Package.name + Package.version) } else if (parseInt(version) > parseInt(Package.version)) { console.error(Config.err.encryptOld + Package.name) } else if (parseInt(version) < parseInt(Package.version)) { console.error(Config.err.encryptNew + Package.name) }

    // 00  01  02  03  04  05  06  07  08  09  10  11  12  13  14  16  17  18
    // W   C   R   Y   P   T   m   v   .   m   v   .   p   v   2   0   0   0
    // 19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
    // 1   2   8   1   2   8   S   H   A   -   5   1   2   i   n   i   t   v
    // 37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
    // e   c   t   o   r   -   s   a   l   t   -   -   -   -   -   -   -   -
    // 55  56  57  58  59  60  61  62  63  64  65  66
    // -   -   -   -   <   W   c   R   y   P   +   >
    config.derive.iterations = parseInt((data.slice(
      exports.head.iter.b, exports.head.iter.e)).toString('utf8'))
    config.crypto.tagLength = parseInt(data.slice(
      exports.head.tag.b, exports.head.tag.e).toString('utf8'))
    config.derive.length = parseInt(data.slice(
      exports.head.length.b, exports.head.length.e).toString('utf8'))
    config.derive.hash = data.slice(
      exports.head.hash.b, exports.head.hash.e).toString('utf8')
    material.iv = transcoder.buf2ab(
      data.slice(exports.head.iv.b, exports.head.iv.e))
    material.salt = transcoder.buf2ab(
      data.slice(exports.head.salt.b, exports.head.salt.e))

    return {
      config: config,
      material: material,
      payload: data.slice(exports.head.salt.e + (config.delimiter.length))
    }
  },

  setHeader: function (config) {
    const sigPos = (exports.getSignature(config)).length
    const iterPos = sigPos + (config.derive.iterations.toString()).length
    const tagPos = iterPos + (config.crypto.tagLength.toString()).length
    const _length = tagPos + (config.derive.length.toString()).length
    const hashPos = _length + (config.derive.hash).length
    const ivPos = hashPos + config.crypto.ivLength
    const saltPos = ivPos + config.derive.saltLength

    exports.head = {
      hash: { b: _length, e: hashPos },
      iter: { b: sigPos, e: iterPos },
      iv: { b: hashPos, e: ivPos },
      length: { b: tagPos, e: _length },
      pref: (config.signaturePrefix).length,
      salt: { b: ivPos, e: saltPos },
      sig: { b: 0, e: sigPos },
      tag: { b: iterPos, e: tagPos }
    }
  },

  version: Package.version

}
